Add Guidance wrt Labelling to Naming and Rules Best Practices

conallob opened the pull request:
Add Guidance wrt Labelling to Naming and Rules Best Practices to docs/practices/naming.md and docs/practices/rules.md, specifically:

* The primary purposes of `job` and `instance`
* Include WARNINGS about accidentally stripping the `job` label, especially in multi-tenant systems


This Fixes https://github.com/prometheus/docs/issues/2690

conallob committed:
Add Important Labels subsection, with job and instance called out

Signed-off-by: Conall O'Brien <conall@conall.net>

conallob committed:
Add a new section about labels

Signed-off-by: Conall O'Brien <conall@conall.net>

beorn7 commented:
With my current patchy availability, there is little chance I get to this anytime soon. Maybe @juliusv has a qualified opinion here?

conallob commented:
Friendly ping @SuperQ @juliusv

conallob committed:
Fix typo

Signed-off-by: Conall O'Brien <conall@conall.net>

conallob commented:
> Hey @conallob, congrats for the nice PR!
> 
> Just one thing: maybe you could fix the `eaach` typo in
> 
> > * The `job` label is a primary key to differentiate metrics from eaach other.
> 
> There is a small confusion I would love to see fixed, in a paragraph just below one of your edits. It is:

Fixed the typo.
 
> > To keep the operations clean, `_sum` is omitted if there are other operations,
> > as `sum()`.
> 
> I don't understand the `as sum()` part. Like, "`x` is omitted if there are other operations such as `x`"? It doesn't make sense to me, in a very basic way. I know it's out of the scope of this PR, but maybe you could touch it to clarify.

I'm afraid that best practice is unrelated.

It also makes sense as written, once you've written enough rules. It's weighing up the trade-off between tracking the chain of operations across a pipeline of rules vs the rule name growing unwieldy. Many of these best practices trace back to specific philosophies from Prometheus' predecessor.

If you still think it needs a polish, please a separate doc bug.

conallob commented:
Ping @juliusv , since @SuperQ is currently unavailable for life reasons

conallob committed:
Update docs/practices/naming.md

Co-authored-by: Ben Kochie <superq@gmail.com>
Signed-off-by: Conall O'Brien <conall@conall.net>

SuperQ commented on docs/practices/naming.md:
```diff
@@ -80,7 +80,18 @@ the underlying metric type and unit you work with.
 of unit and type information in the metric name will cause certain series to collide (e.g. `process_cpu` for seconds and milliseconds).
 
-## Labels
+### Labels
```
```suggestion
## Labels
```

conallob committed:
Update docs/practices/naming.md

Co-authored-by: Ben Kochie <superq@gmail.com>
Signed-off-by: Conall O'Brien <conall@conall.net>

SuperQ commented on docs/practices/rules.md:
```diff
@@ -19,6 +19,8 @@ This page documents proper naming conventions and aggregation for recording rule
 Keeping the metric name unchanged makes it easy to know what a metric is and
 easy to find in the codebase.
 
+IMPORTANT: `job` label acts as a primary key. It is **strongly** recommended that you use it to scope your PromQL expressions to the system you are monitoring.
```
This is misleading. Prometheus doesn't have the concept of "primary key". Not even metric names are a "primary key".

conallob replied:
Fair, especially since folks used to SQL DBs will jump to the conclusion that it's a SQL DB, which it isn't.

Iterated on the language to avoid creating ambiguity

conallob replied:
After noodling on this further, I think the message is still important, but to avoid any ambiguous terms, let's just go with "scoping", since that is what the `job` label is for.

conallob committed:
Iterate on the job label description

Iterate on the description of the job label, removing "primary key", given it's association with SQL

Signed-off-by: Conall O'Brien <conall@conall.net>

conallob commented:
Friendly ping?

conallob commented:
Friendly, you're not at SRECon EMEA this week, ping?

conallob committed:
Update docs/practices/naming.md

Co-authored-by: Ben Kochie <superq@gmail.com>
Signed-off-by: Conall O'Brien <conall@conall.net>

conallob commented on docs/practices/naming.md:
```diff
@@ -82,6 +82,17 @@ of unit and type information in the metric name will cause certain series to col
 
+* `job`
+   * The `job` label is one of the few ubiquitious labels, set at scrape time, and is used to identify metrics scraped from the same target/exporter.
+   * If not specified in PromQL expressions, they will match unrelated metrics with the same name. This is especially true in a multi system or multi tenant installation
```
It applies to all labels. But `job` and `instance` are two uniform labels found on every metric, including ubiquitous synthetic metrics such as `up`

conallob committed:
Merge branch 'prometheus:main' into main

conallob committed:
Merge branch 'prometheus:main' into main

conallob committed:
Merge branch 'main' into main

conallob commented:
Coming back to this long neglected PR, afaik I can tell, the main issue here is whether to put label best practices in the variable naming best practices or not. I only did that since there is no label best practices section currently.


Would creating a new section for label best practices help to unstick this?

conallob committed:
Split labels from naming.md into a new labels.md

Signed-off-by: Conall O'Brien <conall@conall.net>

conallob commented on docs/practices/naming.md:
```diff
@@ -82,6 +82,17 @@ of unit and type information in the metric name will cause certain series to col
+   * If not specified in PromQL expressions, they will match unrelated metrics with the same name. This is especially true in a multi system or multi tenant installation
+
+WARNING: When using `without`, be careful not to strip out the `job` label accidentally.
+
```
I've kept this warning, but moved it to the new labels.md file

conallob commented on docs/practices/naming.md:
```diff
@@ -82,6 +82,17 @@ of unit and type information in the metric name will cause certain series to col
+WARNING: When using `without`, be careful not to strip out the `job` label accidentally.
+
+* `instance`
+   * The `instance` label will include the `ip:port` what was scraped, providing a crucial breadcrumb for debugging scrape time issues
```
Applied

conallob committed:
Manually apply suggested edits from @SuperQ

Signed-off-by: Conall O'Brien <conall@conall.net>

conallob committed:
Set sort_rank to 2

Signed-off-by: Conall O'Brien <conall@conall.net>

conallob committed:
Rename title to just Labels

Signed-off-by: Conall O'Brien <conall@conall.net>

conallob committed:
Wordsmith the IMPORTANT block to describe job as a scoping label, not a primary key

Signed-off-by: Conall O'Brien <conall@conall.net>

conallob committed:
Integrate PR review suggestion

Signed-off-by: Conall O'Brien <conall@conall.net>

conallob commented on docs/practices/naming.md:
```diff
@@ -82,6 +82,17 @@ of unit and type information in the metric name will cause certain series to col
 ## Labels
 
+* `job`
+   * The `job` label is one of the few ubiquitious labels, set at scrape time, and is used to identify metrics scraped from the same target/exporter.
```
Integrated this suggestion, but this section is now in the new labels.md

conallob reviewed (commented):
Reviving this PR, PTAL @SuperQ 


* I've split the label sections of naming.md out into a new labels.md, as discussed in OOB chat with @SuperQ 
* I've also reworded the line mentioning "primary key" to be about defining scope in PromQL expressions, which is the ultimate goal.
* It looks like I need a maintainer to approve the Netlify CI checks though

conallob committed:
Merge branch 'main' into main

conallob committed:
Merge branch 'main' into main

conallob committed:
Merge branch 'main' into main

conallob committed:
Merge branch 'main' into main

conallob committed:
Merge branch 'main' into main

conallob commented:
@SuperQ Friendly ping?

krajorama commented on docs/practices/labels.md:
```diff
@@ -0,0 +1,41 @@
+## Labels
+
+Prometheus labels can come from both the target and from
+[relabeling in discovery](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config) as well as from the target itself.
```
```suggestion
Prometheus labels can come from both the target itself and from
[relabeling in discovery](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config).
```

krajorama commented on docs/practices/labels.md:
```diff
@@ -0,0 +1,41 @@
+By default Prometheus configures two primary discovery target labels.
+
+- `job`
+  - The `job`  is a default target label set by the scrape configs and is used to identify metrics scraped from the same target/exporter.
```
```suggestion
  - The `job` is a default target label set by the scrape configs and is used to identify metrics scraped by the same scrape config.
```

krajorama commented on docs/practices/labels.md:
```diff
@@ -0,0 +1,41 @@
+
+- `job`
+  - The `job`  is a default target label set by the scrape configs and is used to identify metrics scraped from the same target/exporter.
+  - If not specified in PromQL expressions, they will match unrelated metrics with the same name. This is especially true in a multi system or multi tenant installation
```
It might be the intention of the user to not distinguish between unrelated metrics, for example when aggregating across jobs. So I'd turn this into some positive statement instead (use "specify" instead of "not specified"). Multi system seems vague and I don't think multi tenancy has anything to do with this.

```suggestion
  - If specified in PromQL expressions, they will match metrics scraped by the same job.
```

conallob replied:
> It might be the intention of the user to not distinguish between unrelated metrics, for example when aggregating across jobs. So I'd turn this into some positive statement instead (use "specify" instead of "not specified").

I agree that the user may want to aggregate across all, or a subset of jobs. But to do so, they should be explicit with the `job` label values. But I expect unintended label stripping to vastly outnumber cross-job aggregation use cases.

> Multi system seems vague and I don't think multi tenancy has anything to do with this.

"Multi-tenent systems" may not be the best term, but I'm referring to a Prometheus, run as a platform for multiple teams (e.g by a DevEx or Platform Engineering team), to prevent every team running their own siloed Prometheus stack. In such a setup, all PromQL expressions should be scoped with a `job` label, to ensure the metrics are from the the expected exporters.

Or framed another way, in such a centralised stack, always write `up{job=bla}`, never `up{}`

krajorama commented on docs/practices/labels.md:
```diff
@@ -0,0 +1,41 @@
+WARNING: When using `without`, be careful not to strip out the `job` label accidentally.
+
+- `instance`
+  - The `instance` label will include the `ip:port` what was scraped
```
```suggestion
  - The `instance` label will include the `ip:port` what was scraped, identifying the target instance.
```

krajorama commented on docs/practices/labels.md:
```diff
@@ -0,0 +1,41 @@
+  - The `job`  is a default target label set by the scrape configs and is used to identify metrics scraped from the same target/exporter.
+  - If not specified in PromQL expressions, they will match unrelated metrics with the same name. This is especially true in a multi system or multi tenant installation
+
+WARNING: When using `without`, be careful not to strip out the `job` label accidentally.
```
It might be intentional , so I think this needs to be conditional.

krajorama commented on docs/practices/labels.md:
```diff
@@ -0,0 +1,41 @@
+
+- `instance`
+  - The `instance` label will include the `ip:port` what was scraped
+
```
Don't you need a similar warning for "instance" , depending on usage?

conallob replied:
While `instance` is one of the standard scrape time labels, like `job`, stripping it doesn't have the same blast radius. Stripping `instance` will make metrics hard to debug, but should will still work.


For certain use cases that require using multiple layers of rules (e.g in a multi region, multi layered tree of Prometheus), you may want to strip out `instance` at the higher aggregation layers to manage label cardinality (e.g instance labels make sense to the per region aggregation, but can be problematic if aggregated at the global level)


I've added a warning that stripping `instance` can make it harder to debug scrape time issues with a metric though.

krajorama commented on docs/practices/rules.md:
```diff
@@ -27,6 +30,18 @@ If there is no obvious operation to use, use `sum`.  When taking a ratio by
 
+## Labels
+
+NOTE: Omitting a label in a PromQL expression is the functional equivalent of specifying `label=*`
```
Please use proper sentences with punctuation. And that's not valid PromQL label selector.
 
```suggestion
NOTE: Omitting a label in a PromQL expression is the functional equivalent of specifying `label=~".*"`.
```

krajorama commented on docs/practices/rules.md:
```diff
@@ -27,6 +30,18 @@ If there is no obvious operation to use, use `sum`.  When taking a ratio by
+
+NOTE: Omitting a label in a PromQL expression is the functional equivalent of specifying `label=*`
+
+* In both recorded rules and alerting expressions, always specify a `job` label to prevent expression mismatches from occuring.
```
I think the need to specify `job` is very circumstantial , so again I think it needs to be conditional on what you want to achieve. Also specify `job` is very vague in itself.

conallob replied:
Can you elaborate on why `job` is circumstantial?

Afaik, `job` will always be set on metrics unless it is explicitly stripped away.

krajorama commented on docs/practices/rules.md:
```diff
@@ -27,6 +30,18 @@ If there is no obvious operation to use, use `sum`.  When taking a ratio by
+NOTE: Omitting a label in a PromQL expression is the functional equivalent of specifying `label=*`
+
+* In both recorded rules and alerting expressions, always specify a `job` label to prevent expression mismatches from occuring.
+  This is especially important in multi-tenant systems where the same metric names may be exported by different jobs or the
```
I don't think multi-tenant has anything to do with job and instance labels.

conallob replied:
As above, "Multi-tenent systems" may not be the best term, but I'm referring to a Prometheus, run as a platform for multiple teams (e.g by a DevEx or Platform Engineering team), to prevent every team running their own siloed Prometheus stack. In such a setup, all PromQL expressions should be scoped with a `job` label, to ensure the metrics are from the the expected exporters.

Or framed another way, in such a centralised stack, always write `up{job=bla}`, never `up{}`

krajorama commented on docs/practices/rules.md:
```diff
@@ -40,10 +55,6 @@ Instead keep the metric name without the `_count` or `_sum` suffix and replace
 the `rate` in the operation with `mean`. This represents the average
 observation size over that time period.
 
-* Always specify a `without` clause with the labels you are aggregating away.
```
This seems relevant here, why remove it?

conallob replied:
Because it's guidance wrt labels, not naming, so it's moving to the new labels.md. But I forgot to add it to labels.md

Added now.

krajorama reviewed (changes requested):
Thanks for working on this. I feel these additions are currently too specific for your use cases to be included in the generic documentation.

conallob committed:
Update docs/practices/labels.md

Co-authored-by: George Krajcsovits <krajorama@users.noreply.github.com>
Signed-off-by: Conall O'Brien <conall@conall.net>

conallob committed:
Update docs/practices/labels.md

Co-authored-by: George Krajcsovits <krajorama@users.noreply.github.com>
Signed-off-by: Conall O'Brien <conall@conall.net>

conallob committed:
Update docs/practices/labels.md

Co-authored-by: George Krajcsovits <krajorama@users.noreply.github.com>
Signed-off-by: Conall O'Brien <conall@conall.net>

conallob committed:
Update docs/practices/rules.md

Co-authored-by: George Krajcsovits <krajorama@users.noreply.github.com>
Signed-off-by: Conall O'Brien <conall@conall.net>

conallob committed:
Make it clear stripping job is valid in explicit use cases; Extend WARNING to stripping instance

Signed-off-by: Conall O'Brien <conall@conall.net>

conallob committed:
Move label stripping guidance from rules.md

Signed-off-by: Conall O'Brien <conall@conall.net>

conallob committed:
Merge branch 'main' into main

conallob commented:
@krajorama , @SuperQ : friendly ping? I appreciate you're both busy with your day jobs and other prometheus responsibilities, but ideally I'd like to wrap this PR up before it is open for 12+ months (T-29 days)
